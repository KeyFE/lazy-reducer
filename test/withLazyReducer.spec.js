import { withLazyReducer, lazyReducerEnhancer } from '../lib'
import { createStore, combineReducers } from 'redux'
import renderer from 'react-test-renderer'
import React, { Component } from 'react'
import { connect, Provider } from 'react-redux'

@connect(state => {
    return {
        store: state
    }
})
@withLazyReducer({
    nameX: function(state = null, action) {
        return state
    }
})
class InnerComponent extends Component {
    render() {
        return <div>InnerComponent {JSON.stringify(this.props)}</div>
    }
}

@connect(state => {
    return {
        store: state
    }
})
@withLazyReducer(done => {
    setTimeout(() => {
        done({
            reducerWithFun: (state = 'reducerWithFun', action) => {
                return state
            }
        })
    }, 300)
})
class InnerComponentWithFunc extends Component {
    render() {
        return <div>InnerComponent {JSON.stringify(this.props)}</div>
    }
}

@connect(state => {
    return {
        store: state
    }
})
class DelayContainer extends Component {
    constructor(props) {
        super(props)
        this.state = {}
        setTimeout(() => {
            this.setState({
                show: true
            })
        }, 100)
    }

    render() {
        if (this.state.show) {
            return <InnerComponent />
        } else {
            return <div>DelayContainer {JSON.stringify(this.props)}</div>
        }
    }
}

function waitFor(delay) {
    return new Promise(resolve => {
        setTimeout(resolve, delay)
    })
}

test('withLazyReducer', async () => {
    const syncReducers = {
        a: (state = 'a', action) => {
            return state
        },
        b: (state = 'x', action) => {
            return state
        }
    }
    const store = createStore(combineReducers(syncReducers), {}, lazyReducerEnhancer(syncReducers))
    const App = (
        <Provider store={store}>
            <div>
                <DelayContainer />
                <InnerComponentWithFunc />
            </div>
        </Provider>
    )
    const app = renderer.create(App)
    expect(app.toJSON()).toMatchSnapshot()
    await waitFor(200)
    expect(app.toJSON()).toMatchSnapshot()
    await waitFor(400)
    expect(app.toJSON()).toMatchSnapshot()
    
})
