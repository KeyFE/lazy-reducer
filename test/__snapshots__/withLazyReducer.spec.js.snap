// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withLazyReducer 1`] = `
<div>
  <div>
    DelayContainer 
    {"store":{"a":"a","b":"x"}}
  </div>
</div>
`;

exports[`withLazyReducer 2`] = `
<div>
  <div>
    InnerComponent 
    {"store":{"a":"a","b":"x","nameX":null}}
  </div>
</div>
`;

exports[`withLazyReducer 3`] = `
<div>
  <div>
    InnerComponent 
    {"store":{"a":"a","b":"x","nameX":null,"reducerWithFun":"reducerWithFun"}}
  </div>
  <div>
    InnerComponent 
    {"store":{"a":"a","b":"x","nameX":null,"reducerWithFun":"reducerWithFun"}}
  </div>
</div>
`;
